### PR TITLE
Motion detection

### PR DESCRIPTION
**Pull request by DanielPerssonProos**
Finished the core detection module. Integration with rest of system is needed before full functionality.

### CODE REVIEW

File: ViAn/Analysis/AnalysisController.cpp [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+ * Aborts the analysis
+ */
+void AnalysisController::on_abort() {
+    method->abort_analysis();
+}
+
+/**
+ * @brief AnalysisController::on_progress_update
+ * Sends the progress of the current analysis to the gui
+ * @param progress
```
Comment (svidem) [inline ViAn/Analysis/AnalysisController.cpp]: What kind of progress?

File: ViAn/Filehandler/analysis.h [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+#include <map>
+#include "saveable.h"
+#include "opencv2/core/core.hpp"
+
+enum ANALYSIS_TYPE {MOTION_DETECTION = 0, FACIAL_DETECTION = 1};
+class OOI : Saveable{
+    int frame;
+    std::string file_path;
+    std::pair<int, int> upper_left;
+    std::pair<int, int> lower_right;
```
Comment (svidem) [inline ViAn/Filehandler/analysis.h]: Why are these thing not private nor public?

File: ViAn/Filehandler/filehandler.cpp [hunk 3]
```diff
@@ -115,7 +115,7 @@ void FileHandler::write(QJsonObject &json){
     for (auto it = open_projects.begin();  it != open_projects.end(); it++) {
         QJsonObject json_path;
         ID id = *it;
-        Project* proj = get_project(id);
+        Project* proj = get_project(id);
```
Comment (svidem) [inline ViAn/Filehandler/filehandler.cpp]: weird tab at end of line

File: ViAn/Filehandler/videoproject.cpp [hunk 4]
```diff
@@ -79,6 +87,11 @@ void VideoProject::add_bookmark(Bookmark *bookmark){
     this->bookmarks.push_back(bookmark);
 }
 
+ID VideoProject::add_analysis(Analysis analysis){
```
Comment (svidem) [inline ViAn/Filehandler/videoproject.cpp]: uncommented function

**Comment by PatrikLundgren** [comment]
wohoooo!

File: ViAn/Analysis/MotionDetection.h [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+ * @brief The MotionDetection class
+ * This class is used to detect motion in videos. This
+ * is done using background subtraction with OpenCV.
+ */
+class MotionDetection : public AnalysisMethod {
+public:
+    MotionDetection(std::string source_file);
+
+private:
+    cv::Mat fgMaskMOG2, background, prev_frame, shown_frame, diff_frame, result, diff_prev; //fg mask fg mask generated by MOG2 method
```
Comment (parso619) [inline ViAn/Analysis/MotionDetection.h]: `fgMaskMOG2` snake_case
Also the comment should be fixed.

File: ViAn/Analysis/AnalysisMethod.cpp [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    POI* m_POI = new POI();
+    while(!aborted && capture.read(frame)) {
+        // do frame analysis
+        if (sample_current_frame()) {
+            if (scaling_needed)
+                scale_frame();
+
+            std::vector<OOI> detections = analyse_frame();
+
+            if (detections.empty() && detecting) {
```
Comment (PatrikLundgren) [inline ViAn/Analysis/AnalysisMethod.cpp]: easier to read if cases are commented with what they are for

File: ViAn/Analysis/AnalysisMethod.cpp [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+/**
+ * @brief AnalysisMethod::pause_analysis
+ * Sets the necessary bool to pause an analysis.
+ */
+void AnalysisMethod::pause_analysis() {
+    paused = true;
+}
+
+void AnalysisMethod::calculate_scaling_factor() {
```
Comment (PatrikLundgren) [inline ViAn/Analysis/AnalysisMethod.cpp]: missing comment

File: ViAn/Analysis/AnalysisMethod.cpp [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+    int video_width = capture.get(CV_CAP_PROP_FRAME_WIDTH);
+    int video_height = capture.get(CV_CAP_PROP_FRAME_HEIGHT);
+    float height_ratio = float(FULL_HD_HEIGHT)/float(video_height);
+    float width_ratio = float(FULL_HD_WIDTH)/float(video_width);
+     std::cout << "Original width: " << video_width << ", Original height: " << video_height << std::endl;
```
Comment (PatrikLundgren) [inline ViAn/Analysis/AnalysisMethod.cpp]: should std::cout be here?

File: ViAn/Analysis/AnalysisMethod.cpp [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+    if (height_ratio >= 1 && width_ratio >= 1) return;
+
+    scaling_needed = true;
+    //This statement ensures that the original aspect ratio of the video is kept when scaling
+    if (width_ratio >= height_ratio) {
+        std::cout << "Height bigger" << std::endl;
```
Comment (PatrikLundgren) [inline ViAn/Analysis/AnalysisMethod.cpp]: same, std::cout?

File: ViAn/Analysis/AnalysisMethod.cpp [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+        scaling_ratio = height_ratio;
+        scaled_width = int(video_width * scaling_ratio);
+        scaled_height = FULL_HD_HEIGHT;
+    } else {
+        scaling_ratio = width_ratio;
+        scaled_width = FULL_HD_WIDTH;
+        scaled_height = int(video_height * scaling_ratio);
+    }
+    std::cout << "Width: " << scaled_width << ", Height: " << scaled_height << ", Scaling ratio: " << scaling_ratio << std::endl;
```
Comment (PatrikLundgren) [inline ViAn/Analysis/AnalysisMethod.cpp]: cout

File: ViAn/Analysis/AnalysisMethod.cpp [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+void AnalysisMethod::scale_frame() {
```
Comment (PatrikLundgren) [inline ViAn/Analysis/AnalysisMethod.cpp]: missing comment

File: ViAn/Analysis/MotionDetection.cpp [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+    m_analysis.type = MOTION_DETECTION;
+    capture.open(source_file);
+}
+
+/**
+ * @brief MotionDetection::setup_analysis
+ * Initial setup of the analysis
+ */
+void MotionDetection::setup_analysis(){
+    pMOG2 = cv::createBackgroundSubtractorMOG2(500,50,false);
```
Comment (PatrikLundgren) [inline ViAn/Analysis/MotionDetection.cpp]: what are these magic nums?

File: ViAn/Analysis/MotionDetection.cpp [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+/**
+ * @brief MotionDetection::do_analysis
+ * Motion detection specific code
+ * This function is called from the interface thread loop
+ */
+std::vector<OOI> MotionDetection::analyse_frame(){
```
Comment (PatrikLundgren) [inline ViAn/Analysis/MotionDetection.cpp]: this method seems rather complex, if it should be maintainable, perhaps som clarifying comments about implementation is required

File: ViAn/Analysis/MotionDetection.cpp [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+
+    cv::GaussianBlur(frame, frame, blur_size, 0);
+    pMOG2->apply(frame, fgMaskMOG2,-1);
+    pMOG2->getBackgroundImage(background);
+
+    cv::Mat kernel_ero = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(dilation_degree,dilation_degree));
+    cv::threshold(fgMaskMOG2, fgMaskMOG2, 25, 255, cv::THRESH_BINARY);
+    cv::dilate(fgMaskMOG2, fgMaskMOG2, kernel_ero);
+
+    if (!diff_prev.empty() && current_frame % sample_freq == 0) {
```
Comment (PatrikLundgren) [inline ViAn/Analysis/MotionDetection.cpp]: comment if cases please

File: ViAn/Analysis/MotionDetection.cpp [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+        cv::threshold(diff_frame, diff_frame, 25, 255, cv::THRESH_BINARY);
+        cv::dilate(diff_frame, diff_frame, kernel_ero);
+        cv::bitwise_and(diff_frame, fgMaskMOG2, result);
+    } else if (!diff_prev.empty() && !diff_frame.empty()) {
+        cv::bitwise_and(diff_frame, fgMaskMOG2, result);
+    } else {
+        result = fgMaskMOG2.clone();
+    }
+
+    if (current_frame % sample_freq == 0) {
```
Comment (PatrikLundgren) [inline ViAn/Analysis/MotionDetection.cpp]: same, what case is this?

File: ViAn/Analysis/MotionDetection.h [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+#define MOTIONDETECTION_H
+
+#include "AnalysisMethod.h"
+#include "AnalysisController.h"
+#include <string>
+
+/**
+ * @brief The MotionDetection class
+ * This class is used to detect motion in videos. This
+ * is done using background subtraction with OpenCV.
```
Comment (PatrikLundgren) [inline ViAn/Analysis/MotionDetection.h]: good

File: ViAn/Analysis/MotionDetection.h [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+ */
+class MotionDetection : public AnalysisMethod {
+public:
+    MotionDetection(std::string source_file);
+
+private:
+    cv::Mat fgMaskMOG2, background, prev_frame, shown_frame, diff_frame, result, diff_prev; //fg mask fg mask generated by MOG2 method
+    cv::Ptr<cv::BackgroundSubtractor> pMOG2;
+    cv::Size blur_size = cv::Size(31,31);
```
Comment (PatrikLundgren) [inline ViAn/Analysis/MotionDetection.h]: magic 31?

File: ViAn/Filehandler/analysis.h (hunk 2 above)
Comment (PatrikLundgren) [inline ViAn/Filehandler/analysis.h]: members or functions declared within a class is default private. So they are private

File: ViAn/Filehandler/videoproject.cpp (hunk 4 above)
Comment (DanielPerssonProos) [inline ViAn/Filehandler/videoproject.cpp]: That function was written by @PatrikLundgren, so he should look at that.

File: ViAn/Filehandler/analysis.cpp [hunk 18]
```diff
@@ -0,0 +1,241 @@
...
+ * @param json
+ */
+void POI::read(const QJsonObject& json) {
+    this->start_frame = json["start"].toInt();
+    this->end_frame = json["end"].toInt();    
+    for(int i = start_frame; i != end_frame; i++){        
+        QJsonArray json_frame_OOIs = json[QString::number(i)].toArray();
+        std::vector<OOI> oois;
+        for(int j  = 0; j = json_frame_OOIs.size(); j++){            
+            OOI ooi;
```
Comment (svidem) [inline ViAn/Filehandler/analysis.cpp]: j != json_frame_OOIs.size(), this will break alot of things otherwise

File: ViAn/GUI/mainwindow.cpp [hunk 19]
```diff
@@ -9,13 +9,16 @@
 #include <thread>
 #include "icononbuttonhandler.h"
 #include "Video/shapes/shape.h"
+#include "Analysis/MotionDetection.h"
+#include "Analysis/AnalysisMethod.h"
+#include "Analysis/analysiscontroller.h"
```
Comment (parso619) [inline ViAn/GUI/mainwindow.cpp]: Should be `AnalysisController.h`.

File: ViAn/GUI/mainwindow.h [hunk 20]
```diff
@@ -23,6 +23,7 @@
 #include <QMutex>
 #include <QWaitCondition>
 #include "makeproject.h"
+#include "Analysis/analysiscontroller.h"
```
Comment (parso619) [inline ViAn/GUI/mainwindow.h]: Should be `AnalysisController.h`.

**Pull request merged**